mat_dp_pipeline/sdf/standard_data_format.py

```diff
@@ -117,49 +112,78 @@ def is_leaf(self) -> bool:
         return self.targets is not None
 
     def validate(self) -> None:
-        def has(item) -> bool:
-            return item is not None
+        def validate_yearly_keys(base: pd.DataFrame, yearly: dict[Year, pd.DataFrame]):
+            base_keys = set(base.index.unique())
+
+            for year, df in yearly.items():
+                validate(
+                    set(df.index.unique()) <= base_keys,
+                    f"{self.name}: Yearly file ({year}) introduces new items!",
+                )
 
         validate(
-            has(self.targets) != (len(self.children) > 0),
-            "SDF must either have children in the hierarchy or defined targets (leaf level)",
+            (self.targets is not None) != (len(self.children) > 0),
+            f"{self.name}: SDF must either have children in the hierarchy or defined targets (leaf level)",
         )
         validate(
-            has(self.intensities) or not has(self.intensities_yearly),
-            "No base intensities, while yearly files provided!",
+            (self.intensities is not None) or (self.intensities_yearly is None),
```

Comment:
I maintain this is clearer, as it much more closely matches the message:
```py
if (self.intensities is None) and (self.intensities_yearly is not None):
     raise ValueError("No base indicators, while yearly files provided!")
```
I suppose if you really want validate, perhaps this
```py
validate(not(self.intensities is None) and (self.intensities_yearly is not None),"No base indicators, while yearly files provided!")
```

Reply:
I think these are equivalent, via De Morgan's law, but I don't think A at all clearly implies B:
```py
(self.intensities is not None) or (self.intensities_yearly is None),
```
```py
not((self.intensities is None) and (self.intensities_yearly is not None))
```

Reply:
As below - remove `validate` ?

Reply:
yep